includes/cli/class-woocommerce-subscriptions.php

```diff
@@ -0,0 +1,182 @@
+		WP_CLI::line( '' );
+		if ( ! WooCommerce_Subscriptions_Integration::is_enabled() ) {
+			WP_CLI::error( 'WooCommerce Subscriptions Integration is not enabled.' );
+			WP_CLI::line( '' );
+			return;
+		}
+		self::$ids     = isset( $assoc_args['ids'] ) ? explode( ',', $assoc_args['ids'] ) : false;
+		self::$live    = isset( $assoc_args['live'] ) ? true : false;
+		self::$verbose = isset( $assoc_args['verbose'] ) ? true : false;
+		$updated       = 0;
+		$page          = 1;
+		$per_page      = 25;
+		$subscriptions = self::get_subscriptions( $page );
+		if ( empty( $subscriptions ) ) {
+			WP_CLI::success( 'No on-hold subscriptions to process.' );
+			WP_CLI::line( '' );
+			return;
+		}
+		WP_CLI::line( 'Processing subscriptions in ' . ( self::$live ? 'live' : 'dry run' ) . ' mode...' );
+		WP_CLI::line( '' );
+		while ( ! empty( $subscriptions ) ) {
+			foreach ( $subscriptions as $subscription ) {
+				$id = $subscription->get_id();
+				if ( self::$verbose ) {
+					WP_CLI::line( 'Processing subscription ' . $id . '...' );
+				}
+				// A pending retry indicates the subscription is awaiting payment retry.
+				if ( $subscription->get_date( 'payment_retry' ) > 0 ) {
+					if ( self::$verbose ) {
+						WP_CLI::line( 'Subscription is awaiting payment retry. Moving to next subscription...' );
+						WP_CLI::line( '' );
+					}
+					continue;
+				}
+				$renewal_order = $subscription->get_last_order(
+					'all',
+					[ 'renewal' ],
+					[
+						'completed',
+						'processing',
+						'refunded',
+					]
+				);
+				// No failed or pending renewal orders indicates the subscription was likely manually placed on hold.
+				if ( empty( $renewal_order ) ) {
+					if ( self::$verbose ) {
+						WP_CLI::line( 'Subscription has no pending renewal orders. Moving to next subscription...' );
+						WP_CLI::line( '' );
+					}
+					continue;
+				}
+				$last_retry = \WCS_Retry_Manager::store()->get_last_retry_for_order( wcs_get_objects_property( $renewal_order, 'id' ) );
+				// No retries indicates the subscription was likely manually placed on hold.
+				if ( empty( $last_retry ) ) {
+					if ( self::$verbose ) {
+						WP_CLI::line( 'No retries scheduled. Moving to next subscription...' );
+						WP_CLI::line( '' );
+					}
+					continue;
+				}
+				// A non failed status indicates the retry was either manually cancelled
+				// or was successful at one point but likely placed on hold for some other reason.
+				if ( 'failed' !== $last_retry->get_status() ) {
+					if ( self::$verbose ) {
+						WP_CLI::line( 'Last retry does not have a failed status. Moving to next subscription...' );
+						WP_CLI::line( '' );
+					}
+					continue;
+				} else {
+					if ( self::$verbose ) {
+						WP_CLI::line( 'Updating subscription status to expired...' );
+					}
+					if ( self::$live ) {
+						$subscription->update_status( 'expired', __( 'Subscription status updated by Newspack CLI command.', 'newspack-plugin' ) );
```

Comment:
Let's also add one of those Subscriptions Notes that show up in the subscription page?

Reply:
In discussion with Tracy and Katie, it would be best if the subscription had the expiration date of the last failed payment rather than the expiration date of when the script was run.

Reply:
> Let's also add one of those Subscriptions Notes that show up in the subscription page?

The second argument to `$subscription->update_status` is added as a subscription note. Is this what you were referring to?

 ![Screenshot 2024-12-06 at 11 44 11](https://github.com/user-attachments/assets/db74dae1-906f-43aa-8525-584ef1d47e7c)

> In discussion with Tracy and Katie, it would be best if the subscription had the expiration date of the last failed payment rather than the expiration date of when the script was run.

Sure thing @claudiulodro! I updated to script to use the date of the last retry object. This should be the same as the last failed payment: https://github.com/Automattic/newspack-plugin/pull/3593/commits/b29bdd17b357d01af080d0d7e237201734f9d1fc

When testing though, since we are manually triggering the scheduled actions, this end date will be set as though the scheduled action was not triggered manually, using the expected end date for the retry. I think this should be fine since our publishers shouldn't be doing this, however let me know if we need to be more specific here.